lib/src/local_working_copy.rs

```diff
@@ -571,14 +683,21 @@ impl TreeState {
         store: Arc<Store>,
         working_copy_path: PathBuf,
         state_path: PathBuf,
+        exec_config: Option<bool>,
     ) -> Result<TreeState, TreeStateError> {
-        let mut wc = TreeState::empty(store, working_copy_path, state_path);
+        let mut wc = TreeState::empty(store, working_copy_path, state_path, exec_config);
         wc.save()?;
         Ok(wc)
     }
 
-    fn empty(store: Arc<Store>, working_copy_path: PathBuf, state_path: PathBuf) -> TreeState {
+    fn empty(
+        store: Arc<Store>,
+        working_copy_path: PathBuf,
+        state_path: PathBuf,
+        exec_config: Option<bool>,
```

Comment:
Perhaps, `exec_config` can be carried by `SnapshotOptions`? When materializing working-copy files, it's probably okay to try `chmod()` even if the filesystem doesn't support exec bit.

It's a bit odd that `TreeState::empty()` constructor needs it.

Reply:
I can see how `SnapshotOptions` might work, but I'm not sure that it would preserve the correct behavior when doing a checkout/reset/recover and then calling `LockedWorkingCopy::finish()`, since those save to the repo. If we're ok with sending the permission changes regardless of if they might work then I could see it working, but it feels somewhat 'less correct'? Not sure.

Reply:
(assuming `fchmod()` doesn't fail,) checkout is repo->working-copy operation, so it won't have to read exec bit from disk. `reset()` and `recover()` are similar?

Reply:
Yes, but then its the call to `finish()` afterwards that tries to write to the repo. And `finish` doesn't currently take any other configuration. I think it makes sense to keep the config with the tree state.

Reply:
Just to be celar, `LockedLocalWorkingCopy::finish()` doesn't write commits to repo, but writes working-copy/tree states to disk. It's more or less a serialization of in-memory data.

That said, it might be okay to keep a copy of filesystem capability options in `TreeState`. We already have `symlink_support: bool` (and `store: Arc<Store>` which isn't state data but a shared resource.) If we make `symlink_support` configurable, we'll have to resolve it somewhere in `LocalWorkingCopy`, and pass in to `TreeState` constructor (or methods.)

Reply:
Ah, that makes more sense for`finish`.

How would you feel about some sort of `FileCapability` or `Platform` trait/module? It seems like it effectively already exists for symlink and executable bits in `file_util.rs`, and it would be nice to add helpers like a `from_config()` function.